### ISSUE TITLE
PutInstanceInScope with a name? [Enhancement]

### ISSUE THREAD

**Issue by Tobriand** [z4kn4fein/stashbox#66]
Currently the API for `IDependencyResolver.PutInstanceInScope` only allows overriding purely type-based resolutions. If you want to override (or provide for the first time) a dependency under a name, you need access to the actual container. This is fine if you're passing around the `IStashboxContainer`, and can start your scope equivalent via `CreateChildContainer`. However, there's a lot that goes with passing around a container vs a resolver in terms of opening the door to consumers reconfiguring it wholesale in unexpected ways.

Can an overload be added that overrides / provides a named dependency as well on the aforementioned `IDependencyResolver.PutInstanceInScope` method?

**Comment by z4kn4fein** [comment]
Hey,
Yep, sure thing! It'll be available in the next version.

**Comment by Tobriand** [comment]
Thanks!!!

**Comment by z4kn4fein** [comment]
This feature is available now in the latest [pre-release package](https://www.myget.org/feed/pcsajtai/package/nuget/Stashbox), could you please check that it works as expected?

**Comment by Tobriand** [comment]
Hmm... so when I run tests using the pre-release a whole bunch of things fail, but I've as yet been unable to confirm whether it's because of the changes you've put in breaking something or whether I haven't adapted my code to make use of it properly.

One thing I have noticed is there are scenarios where with an **unnamed** instance in scope, calling `scope.GetScopedInstanceOrDefault(the_type)` returns null, despite it being true that accessing the internal `Stashbox.Utils.AvlTreeKeyValue` on the scope can calling `GetOrDefault(the_type)` returning an appropriate value node, which I'm then able to retrieve the value for (by reflection).

This is currently what's breaking my existing tests - in all likelihood because the result is that the scoped value is to all intents and purposes hidden from resolution. I haven't yet looked into why this might be though.

That said, calling `Resolve` on the scope immediately after the item has been put into it works fine, so I'm not sure what's going wrong in the interrim!

I'll post back as and when I've done a more thorough check to see what's going on.

**Comment by z4kn4fein** [comment]
Whoopsie, that doesn't sound good :)
Could you give me an example test case where this issue occurs? Thanks!

**Comment by Tobriand** [comment]
Ok, worked it out. Remember turns out the signature change is potentially breaking:

**Before:**
`IDependencyResolver PutInstanceInScope(Type typeFrom, object instance, bool withoutDisposalTracking = false);`

**Now:**
`IDependencyResolver PutInstanceInScope(Type typeFrom, object instance, object name = null, bool withoutDisposalTracking = false);`

**My use case:**

```
namespace Proxies {
class DependencyResolverProxy : Proxies.IDependencyResolver {

...

public Proxies.IDependencyResolver PutInstanceInScope(Type typeFrom, object instance, bool withoutDisposalTracking = false)
        {
            _Resolver.PutInstanceInScope(typeFrom, instance, withoutDisposalTracking);
            return this;
        }

public Proxies.IDependencyResolver PutInstanceInScope(Type typeFrom, object instance, bool object name, withoutDisposalTracking = false)
        {
            // Awaiting update to StashBox
            //_Resolver.PutInstanceInScope(typeFrom, instance, name, withoutDisposalTracking);
            return this;
        }


...


}
}
```
Which means now the disposal tracking boolean got passed as the name when I didn't have a name argument.

Simplest solution (other than just expecting this to be an edge case) would be to place the optional name argument after the disposal tracking, though that would be inconsistent with the signature elsewhere, or add an overload. Or ignore it; I imagine my implementation is probably not that common (since most use-cases are unlikely to require the additional level of abstraction from the framework).

I can (and have), of course, fixed in my code to explicitly pass the optional argument to the appropriate place.

**Comment by z4kn4fein** [comment]
Yeah, I think you are right, it breaks the current usage because of the object type and also confusing, I changed the API as you mentioned, the name is the last optional parameter now.

**Comment by Tobriand** [comment]
Just to confirm, I've updated to the latest pre-release (500), on which this behavior now works as expected across all scenarios.

**Comment by z4kn4fein** [comment]
released in [v2.8.6](https://github.com/z4kn4fein/stashbox/releases/tag/2.8.6)
